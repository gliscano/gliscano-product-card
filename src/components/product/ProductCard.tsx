/* React */
import React, { createContext, CSSProperties } from 'react';
import { useProduct } from '../../hooks/useProduct';
/* interfaces */
import { OnChangeArgs, Product, ProductCardHandlers, ProductContextProps } from '../../interfaces';
/* Styles */
import styles from '../../styles/styles.module.css';
import '../../styles/custom-styles.css';
import { InitialValues } from '../../interfaces/index';

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

interface ProductCardProps {
  // children?: ReactElement | ReactElement[];
  children?: (arg: ProductCardHandlers) => JSX.Element;
  className?: string;
  product: Product;
  style?: CSSProperties;
  onChange?: (arg: OnChangeArgs) => void;
  value?: number;
  initialValues?: InitialValues
}

const ProductCard = ({
  children,
  className,
  initialValues,
  product,
  style,
  value,
  onChange,
}: ProductCardProps) => {
  const { counter, isMaxCounter, increaseBy, reset } = useProduct({ onChange, product, value, initialValues });

  return (
    <Provider value={{
      counter,
      increaseBy,
      product
    }}>
      <div
        className={`${styles.productCard} ${className}`}
        style={style}
      >
        { children && children({
            count: counter,
            isMaxCountReached: isMaxCounter,
            maxCount: initialValues?.maxCount,
            product,
            increaseBy,
            reset,
          })
        }
      </div>
    </Provider>
  );
};

export default ProductCard;