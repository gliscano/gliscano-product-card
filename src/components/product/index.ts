export { ProductButtons } from './productButtons';
export { ProductImage } from './ProductImage';
export { ProductTitle } from './ProductTitle';
