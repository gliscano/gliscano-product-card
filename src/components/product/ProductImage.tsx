// React
import React, { useContext, CSSProperties } from 'react';
// Context
import { ProductContext } from './ProductCard';
// Styles
import styles from '../../styles/styles.module.css';
// images
import noImage from '../../assets/no-image.jpg';

interface Props {
  img?: string,
  className?: string
  style?: CSSProperties
}

export const ProductImage = ({img, className, style}: Props) => {
    const { product } = useContext( ProductContext );
    let imgToShow: string;
  
    if (img) {
      imgToShow = img;
    } else if (product.img) {
      imgToShow = product.img;
    } else {
      imgToShow = noImage;
    }
  
    return (
      <img
        className={`${styles.productImg} ${className} `}
        style={style}
        src={imgToShow}
        alt="Product"
      />
    )
  };