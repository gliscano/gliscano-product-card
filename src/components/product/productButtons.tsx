// React
import React, { CSSProperties, useContext } from 'react';
// Context
import { ProductContext } from './ProductCard';
// Styles
import styles from '../../styles/styles.module.css';
import '../../styles/custom-styles.css';

interface Props {
  className?: string;
  style?: CSSProperties;
  isMaxCounter?: boolean;
}

export const ProductButtons = ({ className, style, isMaxCounter }: Props) => {
    const { counter, increaseBy } = useContext( ProductContext );
  
    return (
      <div
        className={`${styles.buttonsContainer} ${className} `}
        style={style}
      >
        <button
          className={styles.buttonMinus}
          onClick={() => increaseBy(-1)}
        >
          -
        </button>
        <div className={styles.countLabel}> {counter} </div>
        <button
          className={`${styles.buttonAdd} ${isMaxCounter && styles.disable}`}
          onClick={() => increaseBy(+1)}
          disabled={isMaxCounter}
        >
          +
        </button>
      </div>
    );
  };