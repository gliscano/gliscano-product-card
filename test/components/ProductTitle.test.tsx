import * as React from 'react';
import renderer from 'react-test-renderer'
import ProductCard from '../../src/components/product/ProductCard';
import { ProductTitle } from '../../src/components/product/ProductTitle';
import { product1 } from '../data/products';

describe('Product Title', () => {
  test('Display Component with custom title', () => {
      const wrapper = renderer.create(
        <ProductTitle
            title="Custom Title"
            className='custom-class'
        />
      )
      expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test ('Display component with the product name', () => {
    const wrapper = renderer.create(
      <ProductCard product={product1}>
        {
          () => (
            <ProductTitle />
          )
        }
      </ProductCard>
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  })
});