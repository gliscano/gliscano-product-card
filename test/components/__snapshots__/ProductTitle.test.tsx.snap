// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Title Display Component with custom title 1`] = `
<span
  className="productDescription custom-class"
>
  Custom Title
</span>
`;

exports[`Product Title Display component with the product name 1`] = `
<div
  className="productCard undefined"
>
  <span
    className="productDescription undefined"
  >
    Coffee Mug - 001
  </span>
</div>
`;
